### ISSUE TITLE
`ParserOptions` in `@typescript-eslint/types` is not up-to-date.

### ISSUE THREAD

**Issue by mihe** [typescript-eslint/typescript-eslint#4056]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

I realize this is a niche use-case at best, but when using your `@typescript-eslint/parser` package programmatically I found that the typings for [`ParserOptions`](https://github.com/typescript-eslint/typescript-eslint/blob/9c8f9a89c098e0ebb7bebf88a4a0f07165275fa6/packages/types/src/parser-options.ts#L28-L59) in `@typescript-eslint/types` don't (but probably should) line up with the options listed in [`README.md`](https://github.com/typescript-eslint/typescript-eslint/blob/9c8f9a89c098e0ebb7bebf88a4a0f07165275fa6/packages/parser/README.md) for `@typescript-eslint/parser` as well as with [`TSESTreeOptions`](https://github.com/typescript-eslint/typescript-eslint/blob/9c8f9a89c098e0ebb7bebf88a4a0f07165275fa6/packages/typescript-estree/src/parser-options.ts#L37-L201) in `@typescript-eslint/typescript-estree`.

Here are the discrepancies I noticed:
1. `programs` (plural) is incorrectly called `program` (singular) in `ParserOptions`. It also has the type `Program` instead of `Program[]`. Likely related to #3508.
1. `projectFolderIgnoreList` is incorrectly typed as `(string | RegExp)[]` in `ParserOptions` but was changed to just `string[]` in `TSESTreeOptions`. Likely related to #2418.
1. `preserveNodeMaps` is present in `TSESTreeOptions` but not in `ParserOptions` (intentionally?).
1. `loggerFn` is present in `TSESTreeOptions` but not in `ParserOptions` (intentionally?).
1. `allowAutomaticSingleRunInference` is present in `TSESTreeOptions` but not in `ParserOptions` (intentionally?).

**Repro**

```ts
import { parseForESLint } from '@typescript-eslint/parser';
parseForESLint('', { programs: [] }); // error TS2345 "Object literal may only specify known properties"
```

**Expected Result**

To be able to use any of the above mentioned properties without compile errors or having to augment the typings myself.

**Actual Result**

Compile error TS2345 ("Object literal may only specify known properties").

**Versions**

| package                     | version  |
| --------------------------- | -------- |
| `@typescript-eslint/parser` | `5.2.0`  |
| `TypeScript`                | `4.4.4`  |
| `ESLint`                    | `8.0.1`  |
| `node`                      | `17.0.0` |
